File: day02/day02p1.hs
inRange :: Int -> Int -> Bool
inRange n1 n2 =
  0 < diff && diff < 4
  where
    diff = abs(n1 - n2)

isCorrect :: Int -> Int -> Bool -> Bool
isCorrect n1 n2 isAscending =
  (inRange n1 n2) && ((n1 < n2) == isAscending)

isSafe' :: [Int] -> Bool -> Bool
isSafe' [_] _ = True
isSafe' (n1:n2:ns) isAscending =
  isCorrect n1 n2 isAscending && isSafe' (n2:ns) isAscending

isSafe :: [Int] -> Bool
isSafe (n1:n2:ns) =
  isSafe' (n1:n2:ns) (n1 < n2)

main :: IO()
main = do
  inp <- getContents
  print $ length (filter isSafe (map  (map (read :: String -> Int) . words) (lines inp)))

File: day02/day02p1.py
import itertools
import sys


def in_range(n1: int, n2: int) -> bool:
    return 0 < abs(n1 - n2) < 4


def is_safe(nums: list[int]) -> bool:
    is_ascending = nums[1] > nums[0]
    return all(
        in_range(n1, n2) and (n2 > n1) == is_ascending
        for n1, n2 in itertools.pairwise(nums)
    )


print(sum(is_safe([int(n) for n in line.split()]) for line in sys.stdin))
